Provide an option to log unexpected exceptions from pudb to stderr

mvanderkamp opened the pull request:
It's not much, but it's a start towards making weird issues that are currently swallowed into the abyss a bit easier to debug. The loggers are set up to not write anything out by default, they're just there as something that can be turned on when we want to figure out why something strange is happening.

The logging calls themselves are a bit arbitrary and preliminary in this PR, the main thing I think is having a logging structure in place to begin with for these sorts of situations. Apologies if one already exists that I managed to overlook!

Very much open to feedback on this though, I wasn't sure what direction to go in so this is just a stab in the dark at something that could be useful.

On a related note, I think I'd like to take a pass at cleaning up settings.py, I think there's some odd behaviour in terms of where configs get loaded/saved to, and I think it would be nice if someone using this code as an API could specify a filename for saving/loading configs and/or breakpoint files. But that's for another day...

This is particularly in response to issue #355

mvanderkamp commented:
I should mention- one nice thing about the approach of just logging to stderr is it makes it easy for users to do things like:
```
python -m pudb --log-errors debug_me.py 2> errors.log
```
And then in another shell:
```
tail -f errors.log
```

inducer commented on pudb/var_view.py:
```diff
@@ -456,7 +465,7 @@ def walk_value(self, parent, label, value, id_path=None, attr_prefix=None):
                 key_its.append(dir(value))
             except Exception:
-                pass
+                ui_log.exception('Failed to lookup attributes')
```
```suggestion
                ui_log.exception('Failed to look up attributes')
```

inducer commented on pudb/lowlevel.py:
```diff
@@ -28,6 +28,26 @@
+settings_handler.setFormatter(settings_formatter)
+settings_log = logging.getLogger('settings')
+settings_log.setLevel(logging.CRITICAL)
+settings_log.addHandler(settings_handler)
```
> I should mention- one nice thing about the approach of just logging to stderr is it makes it easy for users to do things like:

That's neat. Could you add a snippet to the docs about that to make that discoverable?

mvanderkamp replied:
Done!

inducer reviewed (commented):
Thanks! This is nice. A few comments below.

asmeurer commented:
Does logging to stderr break the UI? What's the simplest way to test this?

inducer commented:
This LGTM. Thanks for your work on this @mvanderkamp! @asmeurer, any objections?

asmeurer commented:
I didn't get a chance to test this, but as long as it doesn't break the UI it looks fine to me.

mvanderkamp commented:
> Does logging to stderr break the UI? What's the simplest way to test this?

The log does interfere with the UI, which is why I disabled it by default and recommended piping to a file. If you'd prefer, I could require a path argument to the log option and write the logs directly to that file instead. I'm not so familiar with how the various file descriptors are handled here yet, so I wasn't sure of a better approach.

A simple way to try it out is to put something that will raise an exception into a `safely_stringify_for_pudb` method of whatever you're debugging.

asmeurer commented:
What about disabling the UI like it does when stepping through code?

mvanderkamp commented:
Do you mean show it as a popup? Or explicitly show the output stream, require confirmation, then continue?

asmeurer commented:
Right now when you press `n` or `c`, it disables the UI so that any output will just go to the normal terminal history. You can also see this history by pressing `!` or `o`. See the `show_output` function in debugger.py.

asmeurer commented:
You wouldn't necessarily even need to show confirmation, although it might not hurt. It just makes it so that you can have the output visible once PuDB exits, and also not break the UI.

mvanderkamp commented:
I'm not sure I'm clear on how this is done yet- do I have access to that `screen` object from outside the DebuggerUI? If so, do you mean just call `screen.clear()` after each log line, or do the start/stop pairing? Or, do you mean pause the event loop (if I read that right, that's what happens on `n` and `c` keypresses, but I could be mistaken). I'll look into urwid a bit more perhaps, curses is generally a bit magical to me still.

mvanderkamp commented:
This should do the trick but if start/stop would be more efficient I can do that instead.

inducer commented on pudb/lowlevel.py:
```diff
@@ -28,6 +28,39 @@
+ui_handler = SafeStreamHandler()
+ui_handler.setFormatter(ui_formatter)
+ui_log = logging.getLogger('ui')
+ui_log.setLevel(logging.CRITICAL)
```
Given fd083e4, does it still make sense to swallow errors by default? (And, similarly, does the `--log-errors` option still make sense?)

mvanderkamp replied:
I see your point! I've removed the option and set the errors to always be shown

inducer commented on pudb/lowlevel.py:
```diff
@@ -28,6 +28,37 @@
+    """
+    def emit(self, record):
+        from pudb import _have_debugger, _get_debugger
+        super(SafeStreamHandler, self).emit(record)
+        if _have_debugger():
+            dbg = _get_debugger()
+            dbg.ui.screen.clear()
```
Does this wait for the user to see the error message? If not, would it make sense to add that? Alternatively, given that we have a way of getting to the `DebuggerUI` object from here, we could simply call `DebuggerUI.message()` with the content of the message.

mvanderkamp replied:
I'd rather not do that, since the log fires quite regularly when there's a problem in the variable view- once per refresh in fact, if memory serves, so requiring user input every time would render the session unusable. We'd have to figure out a way of caching prior messages and only logging them the first time, which could be possible I suppose, but might also make it harder to figure out why a message is appearing, since it won't reappear.

inducer replied:
That's a fair point. As an alternative, we could direct those messages to pudb's built-in console (which we would have to force-show in the event of output to make sure the messages are seen by the user). How does that sound?

mvanderkamp replied:
I could go with that. I'd want to still send them to stderr as well, or perhaps provide an option to send them to a file instead.

inducer replied:
That sounds reasonable. Thanks for being receptive to the suggestion.

inducer commented on pudb/debugger.py:
```diff
@@ -1595,6 +1595,7 @@ def add_cmdline_content(s, attr):
             self.cmdline_list.set_focus_valign("bottom")
             self.cmdline_list.set_focus(len(self.cmdline_contents) - 1,
                     coming_from="above")
+        self.add_cmdline_content = add_cmdline_content
```
This seems like it wants to be a simple method.

inducer replied:
I think it would be useful if `add_cmdline_content` could force-show the shell (at least until the next time it's focused in the case of auto-hiding) to ensure the output is discoverable.

inducer commented on pudb/debugger.py:
```diff
@@ -2613,7 +2614,8 @@ def make_frame_ui(frame_lineno):
                 except Exception:
-                    pass
+                    from pudb.lowlevel import ui_log
+                    ui_log.exception('Failed to determine class name')
```
`class_name` should perhaps be updated to reflect that an error occurred when trying to find it.

inducer commented on pudb/var_view.py:
```diff
@@ -285,15 +287,15 @@ def type_stringifier(value):
             result = value.safely_stringify_for_pudb()
         except Exception:
-            pass
+            ui_log.exception('safely_stringify_for_pudb call failed')
```
Perhaps update `result` to reflect a brief version of the error? (And point to the internal shell for traceback?)

inducer commented on pudb/var_view.py:
```diff
@@ -285,15 +287,15 @@ def type_stringifier(value):
             return text_type(value)
         except Exception:
-            pass
+            ui_log.exception('string safe type stringifier failed')
```
Perhaps update `result` to reflect a brief version of the error? (And point to the internal shell for traceback?)

inducer reviewed (commented):
Thanks for your work on this. I've tested it and I like it a lot. I've left some final comments. Good to merge after these are addressed from my end.

mvanderkamp commented:
Okay I _think_ that addresses your concerns. I wasn't quite sure about the approach for forcing the command line window, but I believe this does what you describe.

inducer commented:
This is lovely! Thanks so much for working on this.
